### PR TITLE
Feature/148

### LINKED ISSUE #148: Warn on missing security group or rule descriptions
Cloned from https://gitlab.com/jhctechnology/aws-cloudformation-utilities/awslint/issues/17 and https://gitlab.com/jhctechnology/aws-cloudformation-utilities/awslint/issues/18

Security groups and security group rules without a description obscure their purpose and may lead to bad practices in ensuring they only allow traffic from the ports and sources/destinations required.

### PR DESCRIPTION
**Pull request by twellspring**
Fixes #148

### CODE REVIEW

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  def rule_type
+    Violation::WARNING
+  end
+
+  def rule_id
+    'W36'
+  end
+
+  def violating_sg_part(sg_part)
```
Comment (erickascic) [inline lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: make all of these violating* methods private

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  def violating_ingress?(cfn_model)
+    violating_ingress = cfn_model.resources_by_type('AWS::EC2::SecurityGroupIngress').select do |standalone_ingress|
+      standalone_ingress&.description.nil?
+    end
+    violating_ingress.map(&:logical_resource_id)
+  end
+
+  def violating_egress?(cfn_model)
+    violating_egress = cfn_model.resources_by_type('AWS::EC2::SecurityGroupEgress').select do |standalone_ingress|
```
Comment (erickascic) [inline lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: copy-pasta - should be |egress|

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  def violating_security_groups?(cfn_model)
+    violating_security_groups = cfn_model.security_groups.select do |security_group|
+      !violating_sg_part(security_group.ingresses).empty? || !violating_sg_part(security_group.egresses).empty?
+    end
+    violating_security_groups.map(&:logical_resource_id)
+  end
+
+  def violating_ingress?(cfn_model)
+    violating_ingress = cfn_model.resources_by_type('AWS::EC2::SecurityGroupIngress').select do |standalone_ingress|
+      standalone_ingress&.description.nil?
```
Comment (erickascic) [inline lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: this safe navigation operator shouldn't be necessary.  if ingress.description is raising an exception, then there is something we have to fix in cfn-model.  the ModelElement.method_missing() should return nil for any property that isn't found.

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    end
+    violating_ingress.map(&:logical_resource_id)
+  end
+
+  def violating_egress?(cfn_model)
+    violating_egress = cfn_model.resources_by_type('AWS::EC2::SecurityGroupEgress').select do |standalone_ingress|
+      standalone_ingress&.description.nil?
```
Comment (erickascic) [inline lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: it's not enough to check nil.  have to check for an empty string after trimming

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    Violation::WARNING
+  end
+
+  def rule_id
+    'W36'
+  end
+
+  def violating_sg_part(sg_part)
+    sg_part.select do |item|
+      item.description.nil? && item.logical_resource_id.nil?
```
Comment (erickascic) [inline lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: what is this checking?  an inline rule won't have a logic resource id.

i was confused at first by the word "part".   Maybe call the method inline_rules_have_missing_description?(rules)

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb (hunk 2 above)
Comment (twellspring) [reply lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: fixed

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb (hunk 1 above)
Comment (twellspring) [reply lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: fixed

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb (hunk 3 above)
Comment (twellspring) [reply lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: hmmm ... it was consistently giving me an error.  But just took it out and it is working now. So I must have been doing something different when I got the error.   

Resolved

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb (hunk 4 above)
Comment (twellspring) [reply lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: fixed

File: lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb (hunk 5 above)
Comment (twellspring) [reply lib/cfn-nag/custom_rules/SecurityGroupRuleDescriptionRule.rb]: Fixed

**Pull request merged**